recipes/LibriSpeech/SSL/train.py

```diff
@@ -0,0 +1,518 @@
+
+    sb.dataio.dataset.set_output_keys(datasets, ["id", "sig"])
+
+    w2v_mask_collate_fn_partial = partial(
+        w2v_mask_collate_fn,
+        get_out_len_fn=get_output_lengths,
+        mask_prob=hparams["mask_prob"],
+        mask_length=hparams["mask_length"],
+        max_dur=hparams["max_sample_dur"],
+    )
+
+    train_sampler = DynamicBatchSampler(
+        train_data,
+        hparams["seconds_per_batch"],
+        num_buckets=hparams["train_num_buckets"],
+        length_func=lambda x: x["duration"],
+        batch_ordering="random",
+    )
+    if torch.distributed.is_initialized():
+        train_sampler = DistributedSamplerWrapper(train_sampler)
+
+    train_loader = SaveableDataLoader(
+        train_data,
+        batch_sampler=train_sampler,
+        collate_fn=w2v_mask_collate_fn_partial,
+        num_workers=6,
+        pin_memory=True,
+    )
+    valid_loader = SaveableDataLoader(
+        valid_data,
+        collate_fn=w2v_mask_collate_fn_partial,
+        num_workers=hparams["test_dataloader_options"]["num_workers"],
+        batch_size=hparams["test_dataloader_options"]["batch_size"],
+        pin_memory=True,
+    )
+
+    return train_loader, valid_loader
+
+
+def w2v_mask_collate_fn(
+    samples_lst, get_out_len_fn, mask_prob, mask_length, max_dur=16.0
+):
+    """ This creates a batch from a list of samples and also creates
+    the boolean mask that will be used to mask the inputs of the latent
+    encoder. To create the mask we need to know the output shape after the
+    latent extractor, therefore the argument `get_out_len_fn`.
+    One could also create masks per sample (when loading the audio file) and
+    then collate them but at that time one doesn't know the length of the
+    shortest sample in the batch (which determines the number of masked frames)
+    so it's better this way.
+    This function also has the feature to split samples longer than `max_dur` into
+    smaller chunks, which helps reduce the memory usage of attention based systems.
+
+    Arguments
+    ---------
+    samples_lst : list
+        List of samples returned by the audio_pipeline.
+    get_out_len_fn : function
+        Function that calculates length of sample after it passes through feature extractor.
+    mask_prob : float
+        Approximate percentage of frames to mask.
+    mask_length : int
+        Number of contiguous frames that will be masked.
+    max_dur : float
+        Samples longer than this will be split so they're shorter.
+    Returns
+    -------
+    wavs_padded : torch.Tensor, shape (B, T)
+        Audio arrays with right-sided padding.
+    wav_lens : torch.Tensor, shape (B,)
+        For each sample the percentage of the array that is not padding.
+    mask : torch.Tensor, shape (B, T)
+        Boolean mask to mask frames.
+    """
+    wav_lst, latent_length_lst = [], []
+    # If the sample is longer than `max_dur` create several shorter samples from it.
```

Comment:
Wait, the fact that it will split the data will affect the batch size, correct ? So it bypasses Dynamic Batching, in the sense that the number of batches won't correspond ?